README.md
# NTP Client [![CircleCI](https://circleci.com/gh/destinationstransfers/ntp.svg?style=svg)](https://circleci.com/gh/destinationstransfers/ntp) [![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

Lightweight (no dependencies) modern Javascript (Node 8 LTS or later) implementation of the NTP Client Protocol. Based on [node-ntp-client](https://github.com/moonpyk/node-ntp-client) and [NTPServer](https://github.com/Grassboy/NTPServer).

We use it in production at <https://transfers.do> for webhooks timestamps checking (like [Stripe](https://github.com/stripe) and so on) where VM time is not reliable enough.

## Usage

Add the module to your project with `npm install @destinationstransfers/ntp`.

```js
const NTPClient = require('@destinationstransfers/ntp)';

const date = await NTPClient.getNetworkTime();
console.log(date) // 2017-09-20T15:29:09.443Z
```

## API

`getNetworkTime` accepts hash of parameters, below they are with default values:

```js
const { getNetworkTime } = require('@destinationstransfers/ntp)';

const date = await getNetworkTime({
  timeout = 10000, // timeout in ms, default is 10sec
  server = 'time.google.com', // ntp server address
  port = 123, // NTP server port
})
```

You also can override default values using env variables `NPT_PORT`, `NTP_SERVER` and `NTP_REPLY_TIMEOUT`.


CHANGELOG.md
## Version 1.1 - Nov 16, 2017

-   Switched default NPT server to `time.google.com` for better reliability

## Version 1.0 - Nov 13, 2017

-   Switched back to plain Javascript (ES6) for Node 8.x LTS and later
-   Changed API to be `async` / `Promise` based
-   Refactored tests to Jest
-   Simplified payload parsing using `readUInt32BE` (inspired by <https://github.com/Grassboy/NTPServer/blob/master/timeserver.node.js>)
-   Impoved errors handling and increases test coverage


index.test.js
'use strict';

const dgram = require('dgram');

const { getNetworkTime, NTP_REPLY_TIMEOUT } = require('./');

describe('getNetworkTime', () => {
  beforeAll(() => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = NTP_REPLY_TIMEOUT + 1000;
  });

  test('displays the current time', async () => {
    const res = await getNetworkTime();
    expect(res).toBeInstanceOf(Date);
  });

  test('works with another NTP server', async () => {
    const res = await getNetworkTime({ server: 'de.pool.ntp.org' });
    expect(res).toBeInstanceOf(Date);
  });

  test("won't work with an invalid NTP server", async () => {
    await expect(
      getNetworkTime({ server: 'google.com', timeout: 4000 }),
    ).rejects.toBeDefined();
  });

  test('should handle server errors', done => {
    // creating a server
    const port = 41234;
    const server = dgram.createSocket('udp4');

    server.on('error', err => {
      server.close();
      throw err;
    });

    server.once('message', (msg, rinfo) => {
      console.log(new Date());
      console.log(['  message from ', rinfo.address, ':', rinfo.port].join(''));
      if (rinfo.address) {
        // time sync request from client
        // simulate broken response
        server.send(Buffer.alloc(10), 0, 10, rinfo.port, rinfo.address, err => {
          if (err) throw err;
        });
      }
    });

    server.on('listening', async () => {
      const address = server.address();
      // send request
      await expect(
        getNetworkTime({ server: address, port }),
      ).rejects.toBeDefined();
      server.close();
      return done();
    });

    server.bind(port);
  });
});


index.js
'use strict';

const { createSocket } = require('dgram');

/**
 * Amount of acceptable time to await for a response from the remote server.
 * Configured default to 10 seconds.
 */
const {
  NPT_PORT = 123,
  NTP_SERVER = 'time.google.com',
  NTP_REPLY_TIMEOUT = 10 * 1000,
} = process.env;

exports.NTP_REPLY_TIMEOUT = NTP_REPLY_TIMEOUT;

const NPT_MESSAGE_LENGTH = 48;

/**
 * Fetches the current NTP Time from the given server and port.
 *
 * @param {{ntpReplyTimeout: number, server: string, port: number}} opts
 * @returns {Promise<Date>}
 */
function getNetworkTime(
  { timeout = NTP_REPLY_TIMEOUT, server = NTP_SERVER, port = NPT_PORT } = {},
) {
  return new Promise((resolve, reject) => {
    const client = createSocket('udp4');
    const ntpData = Buffer.alloc(NPT_MESSAGE_LENGTH);

    // RFC 2030 -> LI = 0 (no warning, 2 bits), VN = 3 (IPv4 only, 3 bits), Mode = 3 (Client Mode, 3 bits) -> 1 byte
    // -> rtol(LI, 6) ^ rotl(VN, 3) ^ rotl(Mode, 0)
    // -> = 0x00 ^ 0x18 ^ 0x03
    ntpData[0] = 0x1b;

    const tm = setTimeout(() => {
      client.close();
      reject(
        new Error(`Timeout waiting for NTP response from ${server}:${port}`),
      );
    }, timeout);

    /*
      * Some errors can happen before/after send() or cause send() to break.
      * Some errors will also be given to send()
      * NOTE: the error rejection is not generalised, as the client has to
      * lose the connection also, apparently.
    */
    let errorFired = false;

    client.on('error', err => {
      if (errorFired) return;

      errorFired = true;
      clearTimeout(tm);
      reject(err);
    });

    client.send(ntpData, 0, ntpData.length, port, server, err => {
      if (err) {
        if (errorFired) return;
        clearTimeout(tm);
        errorFired = true;
        client.close();
        reject(err);
        return;
      }

      client.once('message', msg => {
        clearTimeout(tm);
        client.close();
        if (msg.length < NPT_MESSAGE_LENGTH) {
          errorFired = true;
          reject(
            new RangeError(
              `Received NTP response from ${server}:${port} is too short: ${
                msg.length
              }`,
            ),
          );
          return;
        }

        // thats inspired by https://github.com/Grassboy/NTPServer/blob/master/timeserver.node.js
        const intpart = msg.readUInt32BE(msg.length - 16);
        const fractpart = msg.readUInt32BE(msg.length - 8);
        const milliseconds = intpart * 1000 + fractpart * 1000 / 0x100000000;

        // **UTC** time
        const date = new Date('Jan 01 1900 GMT');
        date.setUTCMilliseconds(date.getUTCMilliseconds() + milliseconds);

        resolve(date);
      });
    });
  });
}
exports.getNetworkTime = getNetworkTime;
